File: api/src/main/java/net/impactdev/gts/api/data/registry/DeserializerRegistry.java
package net.impactdev.gts.api.data.registry;

import com.google.gson.JsonObject;

import java.util.Optional;
import java.util.function.Function;

public interface DeserializerRegistry {

    <T> void registerDeserializer(Class<T> type, int version, Function<JsonObject, T> deserializer);

    <T> Optional<Function<JsonObject, T>> getDeserializer(Class<T> type, int version);

}

File: api/src/main/kotlin/net/impactdev/gts/api/data/registry/DeserializerRegistry.kt
package net.impactdev.gts.api.data.registry

import com.google.gson.JsonObject
import java.util.*
import java.util.function.Function

interface DeserializerRegistry {
    fun <T> registerDeserializer(type: Class<T>?, version: Int, deserializer: Function<JsonObject?, T>?)
    fun <T> getDeserializer(type: Class<T>?, version: Int): Optional<Function<JsonObject?, T>?>?
}